etcd/cluster-etcd-operator.md

```diff
@@ -0,0 +1,331 @@
+
+- pending:
+
+The default status of a pending member is `MemberUnknown` as we don't really know the reason for the notReadyAddress state without further investigation. When `etcd-member` Pod starts the init containers will wait for various observations to continue. Because we are waiting in a not Ready state we can be simply waiting for direction from the operator. To validate this we do an exploration into the value of `Status.State.Waiting.Reason` if this is observed with a value of `CrashLoopBackOff` and  `finishedAt` time is less than the observationDefault (5mins) we observe in `wait.PollImmediate` for the observationDefault looking for an incrementation in `RestartCount`. If this is true then status for the member is `MemberRemove`. If we pass the `CrashLoopBackOff` test the status is `MemberAdd`.
+
+```
+apiVersion: v1
+items:
+- apiVersion: operator.openshift.io/v1
+  kind: Etcd
+  metadata:
+    annotations:
+      release.openshift.io/create-only: "true"
+    creationTimestamp: "2019-10-02T12:38:41Z"
+    generation: 2
+    name: cluster
+    resourceVersion: "108019"
+    selfLink: /apis/operator.openshift.io/v1/etcds/cluster
+    uid: 905931e7-e511-11e9-8e06-020bc572eaf2
+  spec:
+    forceRedeploymentReason: ""
+    logLevel: ""
+    managementState: Managed
+    observedConfig:
+      cluster:
+        members:
+        - name: etcd-member-ip-10-0-154-109.us-east-2.compute.internal
+          peerURLs: https://etcd-0.sbatsche.devcluster.openshift.com:2380
+          status: Ready
+        - name: etcd-member-ip-10-0-164-65.us-east-2.compute.internal
+          peerURLs: https://etcd-2.sbatsche.devcluster.openshift.com:2380
+          status: Ready
+        pending:
+        - name: etcd-member-ip-10-0-135-168.us-east-2.compute.internal
+          peerURLs: https://etcd-1.sbatsche.devcluster.openshift.com:2380
+          status: Remove
+```
+
+#### InitContainers
+
+##### discovery
+
+The discovery container converts observations from the cluster into configurations required to add member to cluster. 
+
+##### certs
+
+certs container reads secrets containing cert data for that member and persists them to disk.
+
+## Design Details
+
+
+### Test Plan
+
+1. Unit tests for all controllers
+1. New e2e suite to exercise scaling operations and failure modes.
+1. QE will be asked to have cluster-etcd-operator enabled in all test clusters, especially long-lived clusters
+1. The OpenShift Online environments will always run with cluster-etcd-operator enabled
+
+
+### Graduation Criteria
+
+##### This feature will GA in 4.3 as long as:
```

Comment:
Is supporting installation in 4.3 but deferring support for (and explicit testing of) day 2 scaling and automated recovery an option worth calling out?